src/Panel/Panel.php

```diff
@@ -109,32 +602,260 @@ public static function render(App $kirby)
+
+            // route needs authentication?
+            $auth   = $route->attributes()['auth'] ?? true;
+            $areaId = $route->attributes()['area'] ?? null;
+            $area   = $areas[$areaId] ?? null;
+
+            // call the route action to check the result
+            try {
+                // check for access before executing area routes
+                if ($auth !== false) {
+                    static::firewall($kirby->user(), $areaId);
+                }
+
+                $result = $route->action()->call($route, ...$route->arguments());
+            } catch (Redirect $e) {
+                $result = Response::redirect($e->location());
+            } catch (Throwable $e) {
+                $result = static::error($kirby, $e->getMessage());
+            }
+
+            // pass responses directly down to the Kirby router
+            if (is_a($result, 'Kirby\Http\Response') === true) {
+                return $result;
+            }
+
+            // interpret strings as errors
+            if (is_string($result) === true) {
+                $result = static::error($kirby, $result);
+            }
+
+            // only expect arrays from here on
+            if (is_array($result) === false) {
+                $result = static::error($kirby, 'Invalid Panel response');
+            }
+
+            // create the view based on the current area
+            $view = static::view($kirby, $area, $result ?? []);
+
+            $kirby->trigger('panel.route:after', compact('route', 'path', 'method', 'result', 'view'));
+
+            return static::render($kirby, $view['component'], [
+                '$view'  => $view,
+                '$areas' => array_map(function ($area) {
+                    // routes should not be included in the frontend object
+                    unset($area['routes']);
+                    return $area;
+                }, $areas)
+            ]);
+        });
+    }
+
+    /**
+     * Extract the routes from the given array
+     * of active areas.
+     *
+     * @return array
+     */
+    public static function routes(App $kirby, array $areas): array
+    {
+        // the browser incompatibility
+        // warning is always needed
+        $routes = [
+            [
+                'pattern' => 'browser',
+                'action'  => function () use ($kirby) {
+                    return new Response(
+                        Tpl::load($kirby->root('kirby') . '/views/browser.php')
+                    );
+                },
             ]
-        ]);
+        ];
+
+        // register all routes from areas
+        foreach ($areas as $areaId => $area) {
+            foreach ($area['routes'] as $route) {
+                $route['area'] = $areaId;
+                $routes[] = $route;
+            }
+        }
+
+        // redirect routes
+        $routes[] = [
+            'pattern' => [
+                '/',
+                'installation',
+                'login',
+            ],
```

Comment:
Is this route acting as a fallback for the defined patterns if the areas that normally define these routes are not active? So basically "if the user is already logged in, but visits `/login`, use this fallback"? That's how I understand it. If that's correct, we should add a comment that explains this behavior so we still know what this code does in the future.

Reply:
You understand it correctly. Let me add a comment.